security/advancedtls/advancedtls.go

```diff
@@ -0,0 +1,375 @@
+	}()
+	select {
+	case err := <-errChannel:
+		if err != nil {
+			return nil, nil, err
+		}
+	case <-ctx.Done():
+		return nil, nil, ctx.Err()
+	}
+	return WrapSyscallConn(rawConn, conn), credentials.TLSInfo{conn.ConnectionState()}, nil
+}
+
+func (c *advancedTLSCreds) ServerHandshake(rawConn net.Conn) (net.Conn,
+	credentials.AuthInfo, error) {
+	cfg := cloneTLSConfig(c.config)
+	// We build server side verification function only when root cert reloading is needed.
+	if c.getRootCAs != nil {
+		cfg.VerifyPeerCertificate = buildVerifyFunc(c, "", rawConn)
+	}
+	conn := tls.Server(rawConn, cfg)
+	if err := conn.Handshake(); err != nil {
+		return nil, nil, err
+	}
+	return WrapSyscallConn(rawConn, conn), credentials.TLSInfo{conn.ConnectionState()}, nil
+}
+
+func (c *advancedTLSCreds) Clone() credentials.TransportCredentials {
+	return &advancedTLSCreds{
+		config:     cloneTLSConfig(c.config),
+		verifyFunc: c.verifyFunc,
+		getRootCAs: c.getRootCAs,
+		isClient:   c.isClient,
+	}
+}
+
+func (c *advancedTLSCreds) OverrideServerName(serverNameOverride string) error {
+	c.config.ServerName = serverNameOverride
+	return nil
+}
+
+// The function buildVerifyFunc is used when users want root cert reloading, and possibly custom
+// server authorization check.
+// We have to build our own verification function here because current tls module:
+// 1. does not have a good support on root cert reloading
+// 2. will ignore basic certificate check when setting InsecureSkipVerify to true
+func buildVerifyFunc(c *advancedTLSCreds,
+	serverName string,
+	rawConn net.Conn) func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+	verifyFunc := func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+		// If users didn't specify either rootCAs or getRootCAs on client side,
+		// as we see some use cases such as https://github.com/grpc/grpc/pull/20530,
+		// instead of failing, we just don't validate the server cert and let
+		// application decide via VerifyPeer
+		if c.isClient && c.config.RootCAs == nil && c.getRootCAs == nil {
+			if c.isClient && c.verifyFunc != nil {
+				return c.verifyFunc(serverName, rawCerts, verifiedChains)
+			}
+		}
+		var rootCAs *x509.CertPool
+		if c.isClient {
+			rootCAs = c.config.RootCAs
+		} else {
+			rootCAs = c.config.ClientCAs
+		}
+		// reload root CA certs
+		if rootCAs == nil && c.getRootCAs != nil {
+			reloadRootCAs, err := c.getRootCAs(rawConn, rawCerts)
+			if err != nil {
+				return err
+			}
+			rootCAs = reloadRootCAs
+		}
+		// verify peers' certificates against RootCAs and get verifiedChains
+		certs := make([]*x509.Certificate, len(rawCerts))
+		for i, asn1Data := range rawCerts {
+			cert, _ := x509.ParseCertificate(asn1Data)
+			certs[i] = cert
+		}
+		opts := x509.VerifyOptions{
+			Roots:         rootCAs,
+			CurrentTime:   time.Now(),
+			Intermediates: x509.NewCertPool(),
+		}
+		if !c.isClient {
+			opts.KeyUsages = []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth}
```

Comment:
Do we need to use `ExtKeyUsageServerAuth` for clients?

Reply:
I am not sure... theoretically we need to, but in the implementation here(https://github.com/golang/go/blob/master/src/crypto/tls/handshake_client.go#L804-L820), ExtKeyUsageServerAuth is not set, while for server side, the equivalent field is set(https://github.com/golang/go/blob/master/src/crypto/tls/handshake_server.go#L753).

I added that for now and let's see :-)

Reply:
SG.